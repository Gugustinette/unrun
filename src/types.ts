export interface Result {
  /**
   * The module that was loaded.
   */
  module: any
}
