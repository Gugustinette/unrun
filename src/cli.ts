import process from 'node:process'
import { cac } from 'cac'
import { version } from '../package.json'
import type { Options } from './options'

const cli = cac('unrun')
cli.help().version(version)

cli
  .command('[...files]', 'Run a given JS/TS file', {
    allowUnknownOptions: true,
  })
  .option('--debug', 'Show debug logs', {
    default: false,
  })
  .action(async (input: string[], options: Options) => {
    // Verify an input was given
    if (input.length === 0) {
      throw new Error('[unrun] No input files provided')
    }

    // Lazy load unrun
    const { unrun } = await import('./index')

    // Run unrun with provided options
    await unrun({
      path: input[0],
      debug: options.debug,
    })
  })

async function runCLI(): Promise<void> {
  cli.parse(process.argv, { run: false })

  try {
    await cli.runMatchedCommand()
  } catch (error) {
    console.error((error as Error).message)
    process.exit(1)
  }
}

runCLI()
