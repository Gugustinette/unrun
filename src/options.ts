import type { InputOptions, OutputOptions } from 'rolldown'

export interface Options {
  /**
   * The path to the file to be imported.
   * @default 'custom.config.ts'
   */
  path?: string

  /**
   * Debug mode.
   * Wether or not to keep temporary files to help with debugging.
   * Temporary files are stored in `node_modules/.unrun/` if possible,
   * otherwise in the OS temporary directory.
   * @default false
   */
  debug?: boolean

  /**
   * The preset to use for output generation.
   * @default 'jiti'
   */
  outputPreset?: 'jiti' | 'none'

  /**
   * Whether to make Rolldown's CommonJS wrappers async-friendly.
   * This is necessary if the code being imported uses top-level await
   * inside a CommonJS module.
   * @default true
   */
  makeCjsWrapperAsyncFriendly?: boolean

  /**
   * Additional rolldown input options. These options will be merged with the
   * defaults provided by unrun, with these options always taking precedence.
   */
  inputOptions?: InputOptions

  /**
   * Additional rolldown output options. These options will be merged with the
   * defaults provided by unrun, with these options always taking precedence.
   */
  outputOptions?: OutputOptions
}

export interface ResolvedOptions {
  /**
   * The path to the file to be imported.
   * @default 'custom.config.ts'
   */
  path: string

  /**
   * Debug mode.
   * Wether or not to keep temporary files to help with debugging.
   * Temporary files are stored in `node_modules/.unrun/` if possible,
   * otherwise in the OS temporary directory.
   * @default false
   */
  debug: boolean

  /**
   * The preset to use for output generation.
   * @default 'jiti'
   */
  outputPreset: 'jiti' | 'none'

  /**
   * Whether to make Rolldown's CommonJS wrappers async-friendly.
   * This is necessary if the code being imported uses top-level await
   * inside a CommonJS module.
   * @default true
   */
  makeCjsWrapperAsyncFriendly: boolean

  /**
   * Additional rolldown input options. These options will be merged with the
   * defaults provided by unrun, with these options always taking precedence.
   */
  inputOptions?: InputOptions

  /**
   * Additional rolldown output options. These options will be merged with the
   * defaults provided by unrun, with these options always taking precedence.
   */
  outputOptions?: OutputOptions
}

export function resolveOptions(options: Options = {}): ResolvedOptions {
  return {
    path: options.path || 'custom.config.ts',
    debug: options.debug || false,
    outputPreset: options.outputPreset || 'jiti',
    makeCjsWrapperAsyncFriendly: options.makeCjsWrapperAsyncFriendly ?? true,
    inputOptions: options.inputOptions,
    outputOptions: options.outputOptions,
  }
}
