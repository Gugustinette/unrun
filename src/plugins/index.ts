export { createConsoleOutputCustomizer } from './console-output-customizer'
export { createImportMetaResolveShim } from './import-meta-resolve'
export { createJsonLoader } from './json-loader'
export { createRequireResolveFix } from './require-resolve-fix'
export { createRequireTypeofFix } from './require-typeof-fix'
export { createSourcePathConstantsPlugin } from './source-path-constants'
