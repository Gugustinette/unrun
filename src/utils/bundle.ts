import fs from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'node:path'
import process from 'node:process'
import { pathToFileURL } from 'node:url'
import {
  rolldown,
  type InputOptions,
  type OutputChunk,
  type OutputOptions,
} from 'rolldown'
import {
  createConsoleOutputCustomizer,
  createImportMetaResolveShim,
  createJsonLoader,
  createRequireResolveFix,
  createRequireTypeofFix,
  createSourcePathConstantsPlugin,
} from '../plugins'
import type { ResolvedOptions } from '../options'

export async function bundle(options: ResolvedOptions): Promise<OutputChunk> {
  // Check if the entry file or any dependencies import reflect-metadata
  // This helps us decide whether to enable emitDecoratorMetadata
  let needsDecoratorMetadata = false
  try {
    const entryContent = fs.readFileSync(options.path, 'utf8')
    if (/import\s+["']reflect-metadata["']/.test(entryContent)) {
      needsDecoratorMetadata = true
    }
  } catch {
    // Ignore file read errors
  }

  // Also check related files in the same directory for reflect-metadata
  if (!needsDecoratorMetadata) {
    try {
      const dir = path.dirname(options.path)
      const files = fs.readdirSync(dir)
      for (const file of files) {
        if (file.endsWith('.ts') || file.endsWith('.tsx')) {
          try {
            const content = fs.readFileSync(path.join(dir, file), 'utf8')
            if (/import\s+["']reflect-metadata["']/.test(content)) {
              needsDecoratorMetadata = true
              break
            }
          } catch {
            // Ignore
          }
        }
      }
    } catch {
      // Ignore
    }
  }

  // Default definitions for __dirname, __filename, and import.meta properties
  const defaultDefine: InputOptions['define'] =
    // When bundle-require preset is enabled, avoid using global defines and let the plugin inject per-module values.
    options.outputPreset === 'bundle-require'
      ? undefined
      : {
          __dirname: JSON.stringify(path.dirname(options.path)),
          __filename: JSON.stringify(options.path),
          'import.meta.url': JSON.stringify(pathToFileURL(options.path).href),
          'import.meta.filename': JSON.stringify(options.path),
          'import.meta.dirname': JSON.stringify(path.dirname(options.path)),
          'import.meta.env': 'process.env',
        }

  // Compose feature-specific plugins
  const defaultPlugins: InputOptions['plugins'] = [
    // Handle JSON very early so entry JSON paths are rewritten to JS
    createJsonLoader(),
    // Inject __dirname/__filename/import.meta shims
    createSourcePathConstantsPlugin(),
    // Inline import.meta.resolve("./foo") to a file:// URL when literal
    createImportMetaResolveShim(),
    // Fix require.resolve calls to use correct base path
    createRequireResolveFix(options),
    // Customize console output for namespace objects
    createConsoleOutputCustomizer(),
    // Fix typeof require in ESM
    createRequireTypeofFix(),
  ]

  // Handle decorator metadata by creating a custom tsconfig if needed
  let tsconfigPath = path.resolve(process.cwd(), 'tsconfig.json')
  if (needsDecoratorMetadata) {
    if (process.env.UNRUN_DEBUG === 'true') {
      console.error(
        '[unrun] Enabling decorator metadata for reflect-metadata support',
      )
    }
    try {
      // Create a temporary tsconfig that enables emitDecoratorMetadata
      const tempTsconfig = path.join(
        tmpdir(),
        `tsconfig.unrun.${Date.now()}.json`,
      )
      const baseTsconfig = fs.existsSync(tsconfigPath)
        ? JSON.parse(fs.readFileSync(tsconfigPath, 'utf8'))
        : { compilerOptions: {} }

      const decoratorTsconfig = {
        ...baseTsconfig,
        compilerOptions: {
          ...baseTsconfig.compilerOptions,
          emitDecoratorMetadata: true,
          experimentalDecorators: true,
        },
      }

      fs.writeFileSync(tempTsconfig, JSON.stringify(decoratorTsconfig, null, 2))
      tsconfigPath = tempTsconfig
      if (process.env.UNRUN_DEBUG === 'true') {
        console.error('[unrun] Created temp tsconfig:', tempTsconfig)
      }
    } catch {
      // Fallback to original tsconfig if temp file creation fails
    }
  }

  // Input options (https://rolldown.rs/reference/config-options#inputoptions)
  const inputOptions: InputOptions = {
    input: options.path,
    // Use Node platform for better Node-compatible resolution & builtins
    // See https://rolldown.rs/guide/in-depth/bundling-cjs#require-external-modules
    platform: 'node' as const,
    // Treat all non-relative and non-absolute imports as external dependencies,
    //   except for package "imports" specifiers (starting with '#') which we resolve ourselves.
    //   This ensures bare deps resolve from host project while allowing #imports mapping.
    external: (id: string) =>
      !id.startsWith('.') && !id.startsWith('/') && !id.startsWith('#'),
    // Keep __dirname/__filename/import.meta.url behavior by default
    define: defaultDefine as any,
    // Compose feature-specific plugins
    plugins: defaultPlugins,
    // Resolve tsconfig.json from cwd if present
    tsconfig: tsconfigPath,
    keepNames: true,
    // Configure JSX support with classic mode for pragma support
    jsx: {
      mode: 'classic',
      factory: 'React.createElement',
      fragment: 'React.Fragment',
    },
    // Finally, apply user-provided overrides
    ...(options.inputOptions ?? {}),
  }

  // Setup bundle
  const bundle = await rolldown(inputOptions)

  // Output options (https://rolldown.rs/reference/config-options#outputoptions)
  const outputOptions: OutputOptions = {
    format: 'esm' as const,
    inlineDynamicImports: true,
    // Apply user-provided overrides last
    ...(options.outputOptions ?? {}),
  }

  // Generate bundle in memory
  const rolldownOutput = await bundle.generate(outputOptions)

  // Clean up temporary tsconfig if we created one
  if (
    needsDecoratorMetadata &&
    tsconfigPath !== path.resolve(process.cwd(), 'tsconfig.json')
  ) {
    try {
      fs.unlinkSync(tsconfigPath)
    } catch {
      // Ignore cleanup errors
    }
  }

  // Verify that the output is not empty
  if (!rolldownOutput.output[0]) {
    throw new Error('[unrun] No output chunk found')
  }
  // Return the output chunk
  return rolldownOutput.output[0]
}
