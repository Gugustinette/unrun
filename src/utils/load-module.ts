import { Buffer } from 'node:buffer'
import crypto from 'node:crypto'
import fs from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'node:path'
import process from 'node:process'
import { pathToFileURL } from 'node:url'

/**
 * Import a JS module from code string.
 * Write ESM code to a temp file (prefer project-local node_modules/.unrun) and import it.
 * Cleans up file unless keepFile is true.
 * @param code - The JavaScript code to be imported as a module.
 * @param opts - Optional parameters.
 * @param opts.filenameHint - A hint for the filename to use when creating the temp file.
 * @param opts.keepFile - If true, the temp file will not be deleted after import.
 * @returns The imported module.
 */
export async function loadModule(
  code: string,
  opts: { filenameHint?: string; keepFile?: boolean } = {},
): Promise<any> {
  let moduleUrl: string | null = null
  const { filenameHint, keepFile } = opts
  try {
    const hash = crypto.createHash('sha1').update(code).digest('hex')
    const fname = `${filenameHint ? `${sanitize(filenameHint)}.` : ''}${hash}.mjs`
    const projectNodeModules = path.join(process.cwd(), 'node_modules')
    const outDir = path.join(projectNodeModules, '.unrun')
    const outFile = path.join(outDir, fname)
    if (!fs.existsSync(outFile)) {
      try {
        fs.mkdirSync(outDir, { recursive: true })
        fs.writeFileSync(outFile, code, 'utf8')
      } catch {
        const fallbackDir = path.join(tmpdir(), 'unrun-cache')
        const fallbackFile = path.join(fallbackDir, fname)
        fs.mkdirSync(fallbackDir, { recursive: true })
        fs.writeFileSync(fallbackFile, code, 'utf8')
        moduleUrl = pathToFileURL(fallbackFile).href
      }
    }
    moduleUrl = moduleUrl || pathToFileURL(outFile).href
  } catch {
    moduleUrl = `data:text/javascript;base64,${Buffer.from(code).toString('base64')}`
  }

  let _module
  try {
    _module = await import(moduleUrl)
  } finally {
    if (!keepFile && moduleUrl.startsWith('file://')) {
      try {
        fs.unlinkSync(new URL(moduleUrl))
      } catch {}
    }
  }
  return _module
}

function sanitize(name: string) {
  // Allow word chars plus dot and dash
  return name.replaceAll(/[^\w.-]/g, '_')
}
