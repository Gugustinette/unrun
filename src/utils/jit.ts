import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import type { ResolvedOptions } from '../options'
import { bundle } from './bundle'
import { loadModule } from './load-module'
import { makeCjsWrapperAsyncFriendly } from './make-cjs-wrapper-async-friendly'

export const jit = async (options: ResolvedOptions): Promise<any> => {
  // Resolve the file path to an absolute path
  const filePath = path.resolve(process.cwd(), options.path)

  // Check if the file exists
  if (!fs.existsSync(filePath)) {
    throw new Error(`[unrun] File not found: ${filePath}`)
  }

  // Bundle the code
  const outputChunk = await bundle(options)

  // Post-process: make CommonJS wrappers async-friendly
  let finalCode = outputChunk.code
  if (options.makeCjsWrapperAsyncFriendly ?? true) {
    finalCode = makeCjsWrapperAsyncFriendly(finalCode)
  }

  // Load the generated module
  let _module
  try {
    _module = await loadModule(finalCode, {
      filenameHint: path.basename(options.path),
      keepFile: process.env.UNRUN_DEBUG === 'true',
    })
  } catch (error) {
    throw new Error(
      `[unrun] Import failed (code length: ${finalCode.length}): ${(error as Error).message}`,
    )
  }

  // Return the module
  return _module
}
