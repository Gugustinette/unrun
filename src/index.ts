import path from 'node:path'
import { resolveOptions, type Options } from './options'
import { jit } from './utils/jit'

export async function unrun(options: Options): Promise<any> {
  // Resolve options
  const resolvedOptions = resolveOptions(options)

  // Load the module using JIT compilation
  const module = await jit(resolvedOptions)

  // If the output preset is 'jiti', mimic jiti's export behavior
  if (resolvedOptions.outputPreset === 'jiti') {
    const ext = path.extname(resolvedOptions.path)
    // Prefer default export if present
    if (module && 'default' in module && module.default !== undefined) {
      return module.default
    }

    // If it's an ESM namespace with no exports, return a plain object like jiti
    if (
      module &&
      typeof module === 'object' &&
      module[Symbol.toStringTag] === 'Module' &&
      Object.keys(module).length === 0
    ) {
      return ext === '.mjs' ? module : {}
    }

    // Special-case JSON top-level configs: if the evaluated module is a plain
    // object with a self-referential `default` property (due to our JSON
    // loader), strip it so callers get a clean object.
    if (
      module &&
      typeof module === 'object' &&
      ext === '.json' &&
      (module as any).default === module
    ) {
      const cloned = { ...(module as any) }
      delete (cloned as any).default
      return cloned
    }
  }

  // Otherwise return the module as-is
  return module
}
