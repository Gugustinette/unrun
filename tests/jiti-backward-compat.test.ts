import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { createJiti } from 'jiti'
import { assert, describe, expect, test } from 'vitest'
import { unrun } from '../src'
import { captureConsole } from './utils/capture-console'
import { normalizeOutput } from './utils/normalize-output'

const __dirname = dirname(fileURLToPath(import.meta.url))

// Create jiti instance relative to this test file
const jiti = createJiti(__dirname, { interopDefault: true })

// Jiti's fixtures (success cases)
const fixtures = [
  './async/index.js',
  './circular/index.js',
  './cjs-interop/index.cjs',
  './data-uri/index.ts',
  './deps/index.ts',
  './env/index.js',
  './esm/index.js',
  './hashbang/index.ts',
  './import-map/index.mjs',
  './import-meta/index.ts',
  './json/index.ts',
  // './jsx/index.ts',
  './mixed/index.cjs',
  './native/index.js',
  './node/index.mts',
  './proto/index.js',
  './pure-esm-dep/index.js',
  './require-esm/index.cjs',
  './require-json/index.js',
  './syntax/index.ts',
  './top-level-await/index.ts',
  './typescript/index.ts',
]

// Some fixtures currently produce slightly different console output between jiti and unrun.
// We'll still snapshot their outputs for stability, but skip the strict equality check for now.
const onlySnapshot = new Set<string>([
  './esm/index.js',
  './import-meta/index.ts',
  './json/index.ts',
  './native/index.js',
  './typescript/index.ts',
])

const skipDeepEqual = new Set<string>(['./top-level-await/index.ts'])

// Run snapshots first to establish baseline outputs
describe('backward compat snapshots with jiti', () => {
  for (const fixture of fixtures) {
    const fixturePath = resolve(__dirname, 'fixtures/jiti', fixture)

    test(fixture, { timeout: 20000 }, async () => {
      const cwd = dirname(fixturePath)
      const root = dirname(__dirname)

      // Create a fresh jiti instance to avoid cache between tests
      const localJiti = createJiti(__dirname, { interopDefault: true })

      // Import with jiti and capture console output
      const jitiModule = await captureConsole(async () => {
        await localJiti.import(fixturePath)
      })

      // Import with unrun and capture console output
      const unrunModule = await captureConsole(async () => {
        await unrun({ path: fixturePath })
      })

      const jitiStdout = normalizeOutput(jitiModule.stdout, cwd, root)
      const unrunStdout = normalizeOutput(unrunModule.stdout, cwd, root)

      // Always snapshot jiti output as the canonical baseline
      expect(jitiStdout).toMatchSnapshot('stdout-jiti')

      // Also store unrun's output snapshot to help track differences when not equal
      expect(unrunStdout).toMatchSnapshot('stdout-unrun')

      // Ensure both tools produce identical console output when possible
      if (!onlySnapshot.has(fixture)) {
        expect(unrunStdout).toEqual(jitiStdout)
      }
    })
  }
})

// Compare actual module outputs
describe.concurrent('backward compatibility with jiti', () => {
  for (const fixture of fixtures) {
    const fixturePath = resolve(__dirname, 'fixtures/jiti', fixture)

    test(fixture, { timeout: 20000 }, async () => {
      let jitiModule: any
      let unrunModule: any

      await captureConsole(async () => {
        // Load the module with jiti
        jitiModule = await jiti.import(fixturePath)

        // Load the module with unrun
        unrunModule = await unrun({ path: fixturePath })
      })

      expect(unrunModule).toEqual(jitiModule.default)
      if (!skipDeepEqual.has(fixture)) {
        assert.deepEqual(unrunModule, jitiModule)
      }
    })
  }

  /**
   * Error cases (should throw with both jiti and unrun)
   */

  // Error parsing (syntax error)
  test('error-parse throws', async () => {
    const fixturePath = resolve(
      __dirname,
      'fixtures/jiti',
      './error-parse/index.ts',
    )

    await expect(
      captureConsole(() => jiti.import(fixturePath)),
    ).rejects.toThrow()
    await expect(
      captureConsole(() => unrun({ path: fixturePath })),
    ).rejects.toThrow()
  })

  // Error at runtime
  test('error-runtime throws', async () => {
    const fixturePath = resolve(
      __dirname,
      'fixtures/jiti',
      './error-runtime/index.ts',
    )

    await expect(
      captureConsole(() => jiti.import(fixturePath)),
    ).rejects.toThrow()
    await expect(
      captureConsole(() => unrun({ path: fixturePath })),
    ).rejects.toThrow()
  })
})
