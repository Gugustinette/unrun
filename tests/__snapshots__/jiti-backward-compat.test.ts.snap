// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`backward compat snapshots with jiti > ./async/index.js > stdout-jiti 1`] = `"works"`;

exports[`backward compat snapshots with jiti > ./async/index.js > stdout-unrun 1`] = `"works"`;

exports[`backward compat snapshots with jiti > ./circular/index.js > stdout-jiti 1`] = `"a b c"`;

exports[`backward compat snapshots with jiti > ./circular/index.js > stdout-unrun 1`] = `"a b c"`;

exports[`backward compat snapshots with jiti > ./cjs-interop/index.cjs > stdout-jiti 1`] = `"CJS function default interop test passed"`;

exports[`backward compat snapshots with jiti > ./cjs-interop/index.cjs > stdout-unrun 1`] = `"CJS function default interop test passed"`;

exports[`backward compat snapshots with jiti > ./data-uri/index.ts > stdout-jiti 1`] = `""`;

exports[`backward compat snapshots with jiti > ./data-uri/index.ts > stdout-unrun 1`] = `""`;

exports[`backward compat snapshots with jiti > ./deps/index.ts > stdout-jiti 1`] = `
"npm:config: true
npm:defu {}
npm:destr true
npm:etag: true
npm:is-installed-globally false
npm:mime: true
npm:typescript: true
npm:moment-timezone true
npm:zod: true"
`;

exports[`backward compat snapshots with jiti > ./deps/index.ts > stdout-unrun 1`] = `
"npm:config: true
npm:defu {}
npm:destr true
npm:etag: true
npm:is-installed-globally false
npm:mime: true
npm:typescript: true
npm:moment-timezone true
npm:zod: true"
`;

exports[`backward compat snapshots with jiti > ./env/index.js > stdout-jiti 1`] = `
"process.env true
process.env.TEST true
process.env?.TEST true
import.meta true
import.meta.env.TEST true
import.meta.env?.TEST true"
`;

exports[`backward compat snapshots with jiti > ./env/index.js > stdout-unrun 1`] = `
"process.env true
process.env.TEST true
process.env?.TEST true
import.meta true
import.meta.env.TEST true
import.meta.env?.TEST true"
`;

exports[`backward compat snapshots with jiti > ./esm/index.js > stdout-jiti 1`] = `
"{ utilsLib: { utils: { a: 'a', default: 'default' }, version: '123' } }
{ utils: { a: 'a', default: 'default' } }
{
  file: '<cwd><path>',
  dir: '<cwd>',
  'import.meta.url': 'file://<cwd><path>'
}"
`;

exports[`backward compat snapshots with jiti > ./esm/index.js > stdout-unrun 1`] = `
"{ utilsLib: { utils: { a: 'a', default: 'default' }, version: '123' } }
{ utils: { a: 'a', default: 'default' } }
{
  file: '<cwd><path>',
  dir: '<cwd>',
  'import.meta.url': 'file://<cwd><path>'
}"
`;

exports[`backward compat snapshots with jiti > ./hashbang/index.ts > stdout-jiti 1`] = `"1"`;

exports[`backward compat snapshots with jiti > ./hashbang/index.ts > stdout-unrun 1`] = `"1"`;

exports[`backward compat snapshots with jiti > ./import-map/index.mjs > stdout-jiti 1`] = `"{ alias: 'alias' }"`;

exports[`backward compat snapshots with jiti > ./import-map/index.mjs > stdout-unrun 1`] = `"{ alias: 'alias' }"`;

exports[`backward compat snapshots with jiti > ./import-meta/index.ts > stdout-jiti 1`] = `
"hello! { hello: 'world' }
file://<cwd><path>
hello! custom
file://<cwd><path>+custom.ts
import.meta.dirname: <cwd>
import.meta.filename: <cwd><path>"
`;

exports[`backward compat snapshots with jiti > ./import-meta/index.ts > stdout-unrun 1`] = `
"hello! { hello: 'world' }
file://<cwd><path>
hello! custom
file://<cwd><path>+custom.ts
import.meta.dirname: <cwd>
import.meta.filename: <cwd><path>"
`;

exports[`backward compat snapshots with jiti > ./json/index.ts > stdout-jiti 1`] = `
"Imported : { test: 123 } .default: { test: 123 }
Imported with assertion : { test: 123 } .default: { test: 123 }
Required : { test: 123 } .default: { test: 123 }
Dynamic Imported : [Object: null prototype] { default: { test: 123 }, test: 123 } .default: { test: 123 }"
`;

exports[`backward compat snapshots with jiti > ./json/index.ts > stdout-unrun 1`] = `
"Imported : { test: 123 } .default: { test: 123 }
Imported with assertion : { test: 123 } .default: { test: 123 }
Required : { test: 123 } .default: { test: 123 }
Dynamic Imported : [Object: null prototype] { default: { test: 123 }, test: 123 } .default: { test: 123 }"
`;

exports[`backward compat snapshots with jiti > ./jsx/index.ts > stdout-jiti 1`] = `
"<h1>Hello, nano-jsx!<<path>>
<h1>Hello, preact!<<path>>
<h1>Hello, react!<<path>>
<h1>Hello, vue!<<path>>"
`;

exports[`backward compat snapshots with jiti > ./jsx/index.ts > stdout-unrun 1`] = `
"<h1>Hello, nano-jsx!<<path>>
<h1>Hello, preact!<<path>>
<h1>Hello, react!<<path>>
<h1>Hello, vue!<<path>>"
`;

exports[`backward compat snapshots with jiti > ./mixed/index.cjs > stdout-jiti 1`] = `"Mixed works for: <root>"`;

exports[`backward compat snapshots with jiti > ./mixed/index.cjs > stdout-unrun 1`] = `"Mixed works for: <root>"`;

exports[`backward compat snapshots with jiti > ./native/index.js > stdout-jiti 1`] = `"[Module: null prototype] { default: { hasRequire: false } }"`;

exports[`backward compat snapshots with jiti > ./native/index.js > stdout-unrun 1`] = `"[Module: null prototype] { default: { hasRequire: false } }"`;

exports[`backward compat snapshots with jiti > ./node/index.mts > stdout-jiti 1`] = `"node:test true"`;

exports[`backward compat snapshots with jiti > ./node/index.mts > stdout-unrun 1`] = `"node:test true"`;

exports[`backward compat snapshots with jiti > ./proto/index.js > stdout-jiti 1`] = `"exists: true"`;

exports[`backward compat snapshots with jiti > ./proto/index.js > stdout-unrun 1`] = `"exists: true"`;

exports[`backward compat snapshots with jiti > ./pure-esm-dep/index.js > stdout-jiti 1`] = `
"Enter Program
Enter VariableDeclaration
Enter VariableDeclarator
Enter Identifier
Enter Literal"
`;

exports[`backward compat snapshots with jiti > ./pure-esm-dep/index.js > stdout-unrun 1`] = `
"Enter Program
Enter VariableDeclaration
Enter VariableDeclarator
Enter Identifier
Enter Literal"
`;

exports[`backward compat snapshots with jiti > ./require-esm/index.cjs > stdout-jiti 1`] = `"Works!"`;

exports[`backward compat snapshots with jiti > ./require-esm/index.cjs > stdout-unrun 1`] = `"Works!"`;

exports[`backward compat snapshots with jiti > ./require-json/index.js > stdout-jiti 1`] = `"{ type: 'commonjs' }"`;

exports[`backward compat snapshots with jiti > ./require-json/index.js > stdout-unrun 1`] = `"{ type: 'commonjs' }"`;

exports[`backward compat snapshots with jiti > ./syntax/index.ts > stdout-jiti 1`] = `
"Optional chaining: undefined
Nullish coalescing: 0
Logical or assignment: 50 title is empty.
Logical nullish assignment: 50 20"
`;

exports[`backward compat snapshots with jiti > ./syntax/index.ts > stdout-unrun 1`] = `
"Optional chaining: undefined
Nullish coalescing: 0
Logical or assignment: 50 title is empty.
Logical nullish assignment: 50 20"
`;

exports[`backward compat snapshots with jiti > ./top-level-await/index.ts > stdout-jiti 1`] = `"async value works from sub module"`;

exports[`backward compat snapshots with jiti > ./top-level-await/index.ts > stdout-unrun 1`] = `"async value works from sub module"`;

exports[`backward compat snapshots with jiti > ./typescript/index.ts > stdout-jiti 1`] = `
"Decorator metadata keys: design:type
Decorator called with 3 arguments.
Decorator called with 3 arguments.
Decorator called with 3 arguments.
Decorator called with 1 arguments.
{
  file: '<cwd><path>',
  dir: '<cwd>',
  resolve: '<cwd><path>'
} [class DecoratedClass]
{
  satisfiesTest: {
    firstTest: { name: 'first', avatar: 'https:<path>' },
    secondTest: { name: 'second', avatar: [Object] },
    normalizeUserEntity: [Function: normalizeUserEntity]
  }
}
child
promise resolved"
`;

exports[`backward compat snapshots with jiti > ./typescript/index.ts > stdout-unrun 1`] = `
"Decorator metadata keys: design:type
Decorator called with 3 arguments.
Decorator called with 3 arguments.
Decorator called with 3 arguments.
Decorator called with 1 arguments.
{
  file: '<cwd><path>',
  dir: '<cwd>',
  resolve: '<cwd><path>'
} [class DecoratedClass]
{
  satisfiesTest: {
    firstTest: { name: 'first', avatar: 'https:<path>' },
    secondTest: { name: 'second', avatar: [Object] },
    normalizeUserEntity: [Function: normalizeUserEntity]
  }
}
child
promise resolved"
`;
